security/pkg/nodeagent/secretfetcher/secretfetcher.go

```diff
@@ -443,6 +450,22 @@ func (sf *SecretFetcher) FindIngressGatewaySecret(key string) (secret model.Secr
 	val, exist := sf.secrets.Load(key)
 	secretFetcherLog.Debugf("load secret %s from secret fetcher: %v", key, exist)
 	if !exist {
+		// Sometimes we see that a secret in installed but not in cache because watcher is in an
+		// obsolete state and wasn't reset promptly. We bail this case out by trying fetching
+		// the secret from API call. Since this is a rare case, to avoid complication, we don't add
+		// the secret back to cache as it is not a normal codepath. When watcher recovers, those secret
```

Comment:
if watcher recover, will it trigger another push ?

Reply:
It could be if the channel is still open. I don't see anything in code that prevents duplicate push. From my testing, I don't see any bad consequence either when I trigger double push intentionally.

Reply:
why not to add secret to cache here (to avoid duplicated push) ?

Reply:
Adding secret to cache needs to change the channel size and is not desired as discussed earlier.

Reply:
if using existing ```UpdateK8sSecret``` to add to cache, it will trigger push which may result in racing from codepath added here, which may result in increase channel size. 

however, simply add to cache(don't use ```UpdateK8sSecret```  or refactor it a bit) shouldn't trigger the race condition, and it avoid unnecessary push

Reply:
Alternative way is to still call scrtAdded() to add the secret to cache, but using a new routine and return nil in this function so that the push would happen after the cache consumes the secret. Then no channel resize is needed.  If watch adds the secret again, we may still have double push though as  scrtAdded() doesn't check if the added secret is exactly the same as before. Is there a quick way to check that?

Reply:
@JimmyCYJ will take care of preventing duplicated push(https://github.com/istio/istio/pull/15851), with that, no need to add to cache here.